src/handlers/ask-gpt.ts

```diff
@@ -0,0 +1,87 @@
+ */
+export async function askQuestion(context: Context, question: string) {
+  if (!question) {
+    throw context.logger.error("No question provided");
+  }
+  const { specAndBodies, streamlinedComments } = await recursivelyFetchLinkedIssues({
+    context,
+    owner: context.payload.repository.owner.login,
+    repo: context.payload.repository.name,
+  });
+  const formattedChat = await formatChatHistory(context, streamlinedComments, specAndBodies);
+  console.log("formattedChat", formattedChat);
+  return await askGpt(context, question, formattedChat);
+}
+
+/**
+ * Asks GPT a question and returns the completions
+ * @param context - The context object containing environment and configuration details
+ * @param question - The question to ask GPT
+ * @param formattedChat - The formatted chat history to provide context to GPT
+ * @returns completions - The completions generated by GPT
+ **/
+export async function askGpt(context: Context, question: string, formattedChat: string[]): Promise<CompletionsType> {
+  const {
+    env: { UBIQUITY_OS_APP_NAME },
+    config: { model, similarityThreshold },
+  } = context;
+  let similarComments: CommentSimilaritySearchResult[] = [];
+  let similarIssues: IssueSimilaritySearchResult[] = [];
+  try {
+    similarComments = (await context.adapters.supabase.comment.findSimilarComments(question, similarityThreshold, "")) || [];
+  } catch (error) {
+    context.logger.error(`Error fetching similar comments: ${(error as Error).message}`);
+  }
+  try {
+    similarIssues = (await context.adapters.supabase.issue.findSimilarIssues(question, similarityThreshold, "")) || [];
+  } catch (error) {
+    context.logger.error(`Error fetching similar issues: ${(error as Error).message}`);
+  }
+  let similarText = similarComments.map((comment: CommentSimilaritySearchResult) => comment.comment_plaintext);
+  similarText.push(...similarIssues.map((issue: IssueSimilaritySearchResult) => issue.issue_plaintext));
+  // Remove Null Results (Private Comments)
+  similarText = similarText.filter((text) => text !== null);
+  formattedChat = formattedChat.filter((text) => text !== null);
+  context.logger.info(formattedChat.join(""));
+  // Optimize the context
+  formattedChat = optimizeContext(formattedChat);
+  // ReRank the results based on the question
+  // const reRankedChat = formattedChat.length > 0 ? await context.adapters.voyage.reranker.reRankResults(formattedChat.filter(text => text !== ""), question, 300) : [];
+  similarText = similarText.filter((text) => text !== "");
+  const rerankedText = similarText.length > 0 ? await context.adapters.voyage.reranker.reRankResults(similarText, question) : [];
+  rerankedText.forEach((text) => removeUnwantedChars(text));
+  return context.adapters.openai.completions.createCompletion(
+    question,
+    model,
+    rerankedText,
+    formattedChat,
+    ["typescript", "github", "cloudflare worker", "actions", "jest", "supabase", "openai"],
```

Comment:
> 	4.	Handling Ground Truths: They are indicating that the system uses “ground truths” — meaning predefined correct examples or comments that the system relies on for determining context. Even if the query (or comment) doesn’t provide enough context, the system tries not to make assumptions. For example, if the query asks about “types” in a code snippet without specifying a language, the system shouldn’t assume it’s referring to Python.

Hard coding these things is the wrong approach then. This needs to be dynamic in a new task.